### PR TITLE
Refactor build_tokenizer to use kwargs syntax and specify name 

### PR DESCRIPTION
**Pull request by j316chuck**
Refactor build tokenizer in `train/train.py` and `eval/eval.py` to use  `**kwargs` syntax so that config's are dictionaries, classes/functions have arguments explicitly stated in their signature, and `build_***` uses **kwargs for initializing classes. 

## Unit tests
All tests with `build_tokenizer` pass same as before. 
- `make EXTRA_ARGS='-k test_model' test`
- `make EXTRA_ARGS='-k test_hf_config' test` 
- `make EXTRA_ARGS='-k test_hf_mpt_gen' test`
- `make EXTRA_ARGS='-k test_dataloader' test`

## Integration Test: 
Same training loss scores before and after change: 

Screenshot 2023-08-30 at 5 36 38 PM

Before: https://wandb.ai/mosaic-ml/chuck-runs/runs/yzeqofin?workspace=user-chuck-tang98
After: https://wandb.ai/mosaic-ml/chuck-runs/runs/7ryl9ccf

## Issues Closed
Chipping away at tech [debt](https://id.atlassian.com/login/authorize?continue=https%3A%2F%2Fmosaicml.atlassian.net%2Fbrowse%2FRESEARCH-717&token=eyJraWQiOiJtaWNyb3Mvc2lnbi1pbi1zZXJ2aWNlL2FlN21rM2I2cGxzOWhtaDkiLCJhbGciOiJSUzI1NiJ9.eyJtYXJrZWRWZXJpZmllZCI6ImZhbHNlIiwibG9naW5UeXBlIjoic2Vzc2lvblJlZnJlc2giLCJpc3MiOiJtaWNyb3Mvc2lnbi1pbi1zZXJ2aWNlIiwidXNlcklkIjoiNzEyMDIwOjVkY2IyNmIzLWYzNjctNDc2Yy1hYTJlLTNjNzg3NDhiZjgyNiIsImlzU2xhY2tBcHBTb3VyY2UiOiJmYWxzZSIsImF1ZCI6Imxpbmstc2lnbmF0dXJlLXZhbGlkYXRvciIsIm5iZiI6MTY5MjEzNTk3MCwic2NvcGUiOiJMb2dpbiIsImV4cCI6MTY5MjEzNjA5MCwiaWF0IjoxNjkyMTM1OTcwLCJqdGkiOiJmNTA2MjY1My05YTZmLTRjNTQtOWViMC0yYjhiOWQ4MjQ2ODMiLCJoYXNoZWRDc3JmVG9rZW4iOiIxMDYzYTZkYjY2NTVmZDMyZGYwMjQ5N2FmZjllZWFmYmFhOTU5NzA4ZDA4ZjdmMzg4MDk4YjBjMGJlNWU3Yjg2In0.abK8iEzpDsAT9ca3zLtWLCgkEiYyz0Byng6_Zl68jhrMkR0_w-EXp8MEkWVrWXQWEdpNVRjyb9rxq32VN-STEdR3Y1wXYRI6-7z-V7Uwpk0ZxRNproba8kthUFtu22atf-II_P_zjvE4abkVaeENXCq3i9v5jX1MDgYlF_9t72BnQ_-rFpOYDGhme8NXu5w2Z2G_M3AhTzGSSxi52_0JcCAHwMN39RWQKQR3gsHPFDPC6MGdjnN_dnNB18iYNLzp2BGGBBobCPgruuvMsu9r2wDOC0W34pdFnBkXvTnoCtVp-3jIKPvVDphkuSYl28Y6-rvck86bq_Skwz76zg6DQg&state=eyJoYXNoZWRDc3JmVG9rZW4iOiJhNDAxMTdhYWZmMjkzOWNjNWI5YTE0ZDRjOTMyN2ZlYjAxNGUyNjg4ZWNhMTkzYzhhZDcyNzA5YzJlMWIzNTExIn0%3D)

### CODE REVIEW

File: llmfoundry/utils/config_utils.py [hunk 1]
```diff
@@ -16,14 +16,24 @@
...
+    depending on the must_exist flag. If the convert flag is set to True, then
+    we will convert the value to a python object using OmegaConf.to_container.
     """
     value = cfg.pop(key, None)
-    if value is not None:
+    if value is not None and convert:
+        if not isinstance(value, DictConfig) and not isinstance(
+                value, ListConfig):
+            raise ValueError(
+                'The key: {key} has a value: {value} that cannot be \
```
Comment (dakinggg) [inline llmfoundry/utils/config_utils.py:32]:
```suggestion
                f'The key: {key} has a value of type {type(value)} that cannot be \
```

File: scripts/train/train.py [hunk 2]
```diff
@@ -104,7 +104,7 @@ def build_composer_model(model_cfg: DictConfig,
 
 
 def build_composer_peft_model(
-        model_cfg: DictConfig, lora_cfg: DictConfig,
+        pretrained_model_name_or_path: str, lora_args: Dict[str, Any],
```
Comment (dakinggg) [inline scripts/train/train.py:107]: can you rebase on main? I think some of the changes in here are already merged via one of your other prs.

**Review by dakinggg — COMMENTED** [review]
basically lgtm. will look again after rebase

File: scripts/train/train.py (hunk 2 above)
Comment (j316chuck) [reply scripts/train/train.py:107]: rebased

File: scripts/eval/eval.py [hunk 3]
```diff
@@ -100,7 +100,13 @@ def evaluate_model(model_cfg: DictConfig, dist_timeout: Union[float, int],
                    icl_subset_num_batches: Optional[int]):
     print(f'Evaluating model: {model_cfg.model_name}', flush=True)
     # Build tokenizer and model
-    tokenizer = build_tokenizer(model_cfg.tokenizer)
+    tokenizer_cfg: Dict[str,
+                        Any] = om.to_container(model_cfg.tokenizer,
+                                               resolve=True)  # type: ignore
+    tokenizer_name = tokenizer_cfg['name']
+    tokenizer_kwargs = tokenizer_cfg.get('kwargs', {})
+    tokenizer: AutoTokenizer = build_tokenizer(tokenizer_name,
```
Comment (dakinggg) [inline scripts/eval/eval.py:108]: The type here should be `PreTrainedTokenizerBase`, right?

Comment (j316chuck) [reply scripts/eval/eval.py]: there's an incompatability here where the `build_icl_data_and_gauntlet` uses tokenizer as an AutoTokenizer so vscode was screaming at me. I just removed the types for simplicity? Please let me know if there's a better way to resolve this.

Comment (dakinggg) [reply scripts/eval/eval.py]: ah, the type in `build_icl_data_and_gauntlet` is probably just wrong. can you try fixing that type to be `PreTrainedTokenizerBase`?

**Review by dakinggg — APPROVED** [review]

File: scripts/eval/eval.py (hunk 3 above)
Comment (j316chuck) [reply scripts/eval/eval.py]: Fixed types
